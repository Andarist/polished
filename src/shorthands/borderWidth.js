// @flow
import { directionalProperty } from '../helpers/directionalProperty'
import polish from '../validation/polish'

/**
 * Shorthand that accepts up to four values, including null to skip a value, and maps them to their respective directions.
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...borderWidth('12px', '24px', '36px', '48px')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${borderWidth('12px', '24px', '36px', '48px')}
 * `
 *
 * // CSS as JS Output
 *
 * div {
 *   'borderTopWidth': '12px',
 *   'borderRightWidth': '24px',
 *   'borderBottomWidth': '36px',
 *   'borderLeftWidth': '48px'
 * }
 */

function borderWidth(...values: Array<?string>) {
  return directionalProperty('borderWidth', ...values)
}

export default polish({
  modulePath: 'shorthands/borderWidth',
  types: [
    { key: 'firstWidth', type: 'cssMeasure' },
    { key: 'secondWidth', type: 'cssMeasure' },
    { key: 'thirdWidth', type: 'cssMeasure' },
    { key: 'fourthWidth', type: 'cssMeasure' },
  ],
})(borderWidth)
