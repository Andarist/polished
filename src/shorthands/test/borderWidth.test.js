// @flow
import borderWidth from '../borderWidth'

describe('borderWidth', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
      log: global.console.log,
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })
  it('properly applies a value when passed only one', () => {
    expect(borderWidth('12px')).toMatchSnapshot()
  })
  it('properly applies values when passed two', () => {
    expect(borderWidth('12px', '24px')).toMatchSnapshot()
  })
  it('properly applies values when passed three', () => {
    expect(borderWidth('12px', '24px', '36px')).toMatchSnapshot()
  })
  it('properly applies values when passed four', () => {
    expect(borderWidth('12px', '24px', '36px', '48px')).toMatchSnapshot()
  })
  it('throws an error for each invalid value', () => {
    borderWidth('top', 'right', 'bottom', 'left')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })
  it('throws a warning when passed too many parameters', () => {
    borderWidth('12px', '24px', '36px', '48px', 'none')
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })
})
