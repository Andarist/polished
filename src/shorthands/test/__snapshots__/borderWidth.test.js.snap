exports[`borderWidth properly applies a value when passed only one 1`] = `
Object {
  "borderBottomWidth": "12px",
  "borderLeftWidth": "12px",
  "borderRightWidth": "12px",
  "borderTopWidth": "12px",
}
`;

exports[`borderWidth properly applies values when passed four 1`] = `
Object {
  "borderBottomWidth": "36px",
  "borderLeftWidth": "48px",
  "borderRightWidth": "24px",
  "borderTopWidth": "12px",
}
`;

exports[`borderWidth properly applies values when passed three 1`] = `
Object {
  "borderBottomWidth": "36px",
  "borderLeftWidth": "24px",
  "borderRightWidth": "24px",
  "borderTopWidth": "12px",
}
`;

exports[`borderWidth properly applies values when passed two 1`] = `
Object {
  "borderBottomWidth": "12px",
  "borderLeftWidth": "24px",
  "borderRightWidth": "24px",
  "borderTopWidth": "12px",
}
`;

exports[`borderWidth throws a warning when passed too many parameters 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- shorthands/borderWidth.js --%c

borderWidth %cexpects a maximum of %c4 parameters%c. However, you passed %c5 parameters%c. This additional parameter was ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderWidth%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`borderWidth throws an error for each invalid value 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderWidth.js --%c

borderWidth %cexpects a %c1st parameter%c(%cfirstWidth%c: %ccssMeasure%c). However, you passed (%c\'top\'%c:%cstring%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderWidth%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderWidth.js --%c

borderWidth %cexpects a %c2nd parameter%c(%csecondWidth%c: %ccssMeasure%c). However, you passed (%c\'right\'%c:%cstring%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderWidth%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderWidth.js --%c

borderWidth %cexpects a %c3rd parameter%c(%cthirdWidth%c: %ccssMeasure%c). However, you passed (%c\'bottom\'%c:%cstring%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderWidth%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderWidth.js --%c

borderWidth %cexpects a %c4th parameter%c(%cfourthWidth%c: %ccssMeasure%c). However, you passed (%c\'left\'%c:%cstring%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderWidth%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
